matron/src/hardware/screen/ssd1322.c

```diff
@@ -0,0 +1,393 @@
+        pthread_mutex_destroy(&lock);
+        gpiod_line_release(gpio_reset);
+        gpiod_line_release(gpio_dc);
+        gpiod_chip_close(gpio_0);
+        close(spidev_fd);
+
+        free(spidev_buffer);
+        spidev_buffer = NULL;
+    }
+}
+
+#ifdef SSD1322_USES_THREAD
+void ssd1322_update(cairo_surface_t * surface, bool surface_may_have_color){
+    display_dirty = true;
+    surface_pointer = surface;
+    should_translate_color = surface_may_have_color;
+}
+
+void ssd1322_refresh(){
+#else
+void ssd1322_update(cairo_surface_t * surface_pointer, bool should_translate_color){
+#endif
+    struct spi_ioc_transfer transfer = {0};
+
+    if( spidev_fd <= 0 ){
+        fprintf(stderr, "%s: spidev not yet opened.\n", __func__);
+        return;
+    }
+
+    write_command_with_data(SSD1322_SET_COLUMN_ADDRESS, 28, 91);
+    write_command_with_data(SSD1322_SET_ROW_ADDRESS, 0, 63);
+    write_command(SSD1322_WRITE_RAM_COMMAND);
+
+    if( should_turn_on ){
+        write_command(SSD1322_SET_DISPLAY_ON);
+        should_turn_on = 0;
+    }
+
+    pthread_mutex_lock(&lock);
+
+    const uint32_t surface_w = cairo_image_surface_get_width(surface_pointer);
+    const uint32_t surface_h = cairo_image_surface_get_height(surface_pointer);
+    cairo_format_t surface_f = cairo_image_surface_get_format(surface_pointer);
+
+    if(surface_w != 128 || surface_h != 64 || surface_f != CAIRO_FORMAT_ARGB32){
+        fprintf(stderr, "%s: %ux%u = invalid surface size\n", __func__, surface_w, surface_h);
+        goto early_return;
+    }
+
+    const uint32_t tx_len = surface_w * surface_h;
+    const uint32_t * data = (const uint32_t *) cairo_image_surface_get_data(surface_pointer);
+
+    if( should_translate_color ){
+        // Preserve luminance of RGB when converting to grayscale. Use the
+        // closest multiple of 16 to the fraction to scale the channels'
+        // grayscale value. Use a multiple of 16 because a 4-bit grayscale
+        // value should fit into the upper nibble of the 8-bit value. The
+        // decimal approximation is out of 256: 80 + 160 + 16 = 256.
+        for( uint32_t i = 0; i < tx_len; i += 8 ){
+            uint8x8x4_t pixel = vld4_u8((const uint8_t *) (data + i));
+            uint16x8_t r = vmull_u8(pixel.val[2], vdup_n_u8( 80)); // R * ~ 0.30
+            uint16x8_t g = vmull_u8(pixel.val[1], vdup_n_u8(160)); // G * ~ 0.59
+            uint16x8_t b = vmull_u8(pixel.val[0], vdup_n_u8( 16)); // B * ~ 0.11
+            vst1_u8(spidev_buffer + i, vaddhn_u16(vaddq_u16(r,g), b));
+        }
+    }
+    else{
+        // If the surface has only been drawn to, we can guarantee that RGB are
+        // all equal values representing a grayscale value. So, we can take any
+        // of those channels arbitrarily.
+        for( uint32_t i = 0; i < tx_len; i += 16 ){
+            const uint8x16x4_t RGB = vld4q_u8((uint8_t *) (data + i));
+            vst1q_u8(spidev_buffer + i, RGB.val[0]);
+        }
+    }
```

Comment:
If the aliasing issue is driver-related, I would expect it is somewhere in here.